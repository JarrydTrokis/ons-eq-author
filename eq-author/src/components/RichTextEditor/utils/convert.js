import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = nodeToFn => {
  const htmlToEntity = (nodeName, ...otherArgs) => {
    const entity = Object.entries(nodeToFn)
      .filter(([name]) => name === nodeName)
      .map(([, fn]) => fn(nodeName, ...otherArgs))
      .find(result => result);

    return entity || null;
  };

  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
