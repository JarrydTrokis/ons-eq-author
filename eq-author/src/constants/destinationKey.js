export const NextPage = "NextPage";
export const Default = "Default";
export const EndOfCurrentSection = "EndOfCurrentSection";

export const destinationKey = {
  [EndOfCurrentSection]: "End of current section",
  [NextPage]: "Next page",
  [Default]: "Select a destination",
};
