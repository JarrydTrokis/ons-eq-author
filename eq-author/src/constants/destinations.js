export const NextPage = "NextPage";
export const Default = "Default";
export const EndOfCurrentSection = "EndOfCurrentSection";

export const logicalDestinations = () => [
  {
    id: NextPage,
    logicalDestination: NextPage,
    displayName: NextPage,
  },
  {
    id: EndOfCurrentSection,
    logicalDestination: EndOfCurrentSection,
    displayName: EndOfCurrentSection,
  },
];
