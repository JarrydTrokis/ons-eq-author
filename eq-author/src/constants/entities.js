export const SECTION = "section";
export const PAGE = "page";
export const QUESTION_CONFIRMATION = "question-confirmation";
export const INTRODUCTION = "introduction";
