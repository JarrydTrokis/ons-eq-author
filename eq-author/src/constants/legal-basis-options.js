const LEGAL_BASIS_OPTIONS = [
  {
    value: "NOTICE_1",
    description: "Section 1 of the Statistics of Trade Act 1947.",
  },
  {
    value: "NOTICE_2",
    description: "Sections 2 and 3 of the Statistics of Trade Act 1947.",
  },
  {
    value: "NOTICE_3",
    description: "Sections 3 and 4 of the Statistics of Trade Act 1947.",
  },
  {
    value: "NOTICE_NI",
    description:
      "Article 5 of the Statistics of Trade and Employment (Northern Ireland) Order 1988.",
  },
  {
    value: "NOTICE_FUELS",
    description:
      "Petroleum Act 1998 and Section 1 of the Statistics of Trade Act 1947.",
  },
  {
    value: "VOLUNTARY",
    description: "Voluntary",
  },
];

export default LEGAL_BASIS_OPTIONS;
