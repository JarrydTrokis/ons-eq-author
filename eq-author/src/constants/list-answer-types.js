export const DRIVING = "driving";
export const ANOTHER = "another";
