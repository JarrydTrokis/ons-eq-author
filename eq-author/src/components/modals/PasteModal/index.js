import React, { useState, useEffect, useCallback } from "react";
import styled from "styled-components";
import { colors } from "constants/theme";
import Modal from "components/modals/Modal";
import ButtonGroup from "components/buttons/ButtonGroup";
import Button from "components/buttons/Button";
import PropTypes from "prop-types";

const StyledModal = styled(Modal)`
  .Modal {
    max-width: 500px;
    border-radius: 0.4rem;
  }
`;

const Title = styled.h2`
  color: ${colors.darkGrey};
`;

const Message = styled.div``;

const ButtonGroupStyled = styled(ButtonGroup).attrs({
  horizontal: true,
  align: "right",
})`
  padding-top: 1.5em;
`;

export const FormatText = (value) => {
  value = value.replace(/\s+/g, " ");
  return value;
};

export const preserveRichFormatting = (text) => {
  // Replace multiple spaces and tabs with a single space
  let formattedText = text.replace(/[ \t]+/g, " ");

  // Split the text into lines
  let lines = formattedText.split(/\r?\n/);

  // Remove leading and trailing spaces from each line and join them back with newline characters
  formattedText = lines.map((line) => line.trim()).join("\n");

  return formattedText;
};

const PasteModal = ({
  isOpen = false,
  onConfirm: handleConfirm,
  onCancel: handleCancel,
}) => {
  useEffect(() => {
    const close = (event) => {
      if (event.key === "Escape") {
        handleCancel();
      }
    };

    window.addEventListener("keydown", close);
    return () => window.removeEventListener("keydown", close);
  });

  return (
    <StyledModal
      isOpen={isOpen}
      onClose={handleCancel}
      data={{ test: "paste-modal" }}
    >
      <Title>Confirm the removal of extra spaces from copied content</Title>
      <Message>
        <p>
          The copied content contains extra spaces at the start of lines of
          text, between words, or at the end of lines of text.
        </p>
        <p>
          Extra spaces need to be removed before this content can be pasted.
          Confirming will remove them automatically, while cancelling will
          prevent pasting.
        </p>
      </Message>
      <ButtonGroupStyled>
        <Button
          variant="secondary"
          onClick={handleCancel}
          data-test="paste-modal-cancel"
        >
          Cancel
        </Button>
        <Button
          variant="primary"
          onClick={handleConfirm}
          data-test="paste-modal-confirm"
          autoFocus
        >
          Confirm
        </Button>
      </ButtonGroupStyled>
    </StyledModal>
  );
};

export const usePasteModal = ({ action, ...props }) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleConfirm = useCallback(() => {
    action();
    setIsOpen(false);
  }, [action]);

  const handleCancel = useCallback(() => setIsOpen(false), []);

  return {
    trigger: useCallback(() => setIsOpen(true), []),
    component: useCallback(
      () => (
        <PasteModal
          {...props}
          isOpen={isOpen}
          onConfirm={handleConfirm}
          onCancel={handleCancel}
        />
      ),
      [props, handleCancel, handleConfirm, isOpen]
    ),
  };
};

PasteModal.propTypes = {
  isOpen: PropTypes.bool,
  onConfirm: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

export default PasteModal;
