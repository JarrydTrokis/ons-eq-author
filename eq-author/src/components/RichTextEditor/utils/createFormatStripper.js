import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig as pipedFilterConfig } from "components/RichTextEditor/entities/PipedValue";
import { filterConfig as linkFilterConfig } from "components/RichTextEditor/LinkPlugin";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" },
};

// Custom filter function to replace multiple spaces with a single space
function replaceMultipleSpaces(contentState) {
  const text = contentState.getText();
  const transformedText = text.replace(/\s+/g, " "); // Replace multiple spaces with a single space
  return contentState.merge({
    text: transformedText,
  });
}

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [pipedFilterConfig, linkFilterConfig],
    maxNesting: 0,
    whitespacedCharacters: [replaceMultipleSpaces],
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return (editorState) => filterEditorState(filterConfiguration, editorState);
}
