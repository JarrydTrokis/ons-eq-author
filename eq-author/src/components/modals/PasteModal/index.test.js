import React from "react";
import { render, fireEvent } from "tests/utils/rtl";
import PasteModal, { preserveRichFormatting } from ".";

import { keyCodes } from "constants/keyCodes";

const { Escape } = keyCodes;

describe("preserveRichFormatting function", () => {
  it("should replace multiple spaces and tabs with a single space", () => {
    const inputText = "   Hello \t\t World    ";
    const expectedOutput = "Hello World";
    const result = preserveRichFormatting(inputText);
    expect(result).toBe(expectedOutput);
  });

  test("removes leading and trailing spaces from each line", () => {
    const inputText = "   Line 1   \n   Line 2   \n   Line 3   ";
    const expectedOutput = "Line 1\nLine 2\nLine 3";
    const result = preserveRichFormatting(inputText);
    expect(result).toBe(expectedOutput);
  });
});

describe("PasteModal", () => {
  let onConfirm, onCancel;
  onConfirm = jest.fn();
  onCancel = jest.fn();
  const renderModal = (props) =>
    render(
      <PasteModal isOpen onConfirm={onConfirm} onCancel={onCancel} {...props} />
    );

  it("should close modal when cancel button is clicked", () => {
    const { getByTestId } = renderModal();
    const cancelButton = getByTestId("paste-modal-cancel");
    fireEvent.click(cancelButton);
    expect(onCancel).toHaveBeenCalledTimes(1);
  });

  it("should close modal when confirm button is clicked", () => {
    const { getByTestId } = renderModal();
    const confirmButton = getByTestId("paste-modal-confirm");
    fireEvent.click(confirmButton);
    expect(onConfirm).toHaveBeenCalledTimes(1);
  });
});

describe("PasteModalEscape", () => {
  let onConfirm, onCancel;
  onConfirm = jest.fn();
  onCancel = jest.fn();
  const renderModal = (props) =>
    render(
      <PasteModal isOpen onConfirm={onConfirm} onCancel={onCancel} {...props} />
    );

  it("should close modal when escape key is pressed", () => {
    const { getByTestId } = renderModal();
    fireEvent.keyDown(getByTestId("paste-modal"), {
      key: Escape,
      code: Escape,
    });
    expect(onCancel).toHaveBeenCalledTimes(1);
  });
});
