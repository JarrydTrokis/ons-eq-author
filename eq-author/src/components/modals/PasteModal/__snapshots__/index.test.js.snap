// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasteModal should render the modal correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c3 {
  list-style: none;
  margin: 0;
  padding: 0;
}

.c1 {
  position: absolute;
  bottom: 0;
  text-align: center;
  min-width: 10em;
  margin-bottom: 0;
  z-index: 99999999;
}

.c2 {
  display: inline-block;
  text-align: initial;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 {
  width: 100%;
  height: 100%;
  overflow-y: auto;
  position: relative;
}

.c7 {
  max-height: 100vh;
}

.c8 {
  float: right;
  font-size: 32px;
  cursor: pointer;
  border: none;
  background: none;
  margin: 0;
  margin-top: 0.1rem;
  margin-right: 0.05em;
}

.c4 .Modal {
  padding: 2em;
  position: absolute;
  background: #FFFFFF;
  box-shadow: 0 2px 12px 0 rgba(0,0,0,0.3);
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
  -webkit-transition: all 100ms ease-in 50ms;
  transition: all 100ms ease-in 50ms;
  opacity: 0;
}

.c4 .Modal:focus {
  outline: 3px solid #3B7A9E;
}

.c4 .Modal--after-open {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
}

.c4 .Modal--before-close {
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
  opacity: 0;
}

.c4 .Overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  -webkit-transition: opacity 100ms ease-out;
  transition: opacity 100ms ease-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  z-index: 9999999;
  background-color: rgba(0,0,0,0.5);
}

.c4 .Overlay--after-open {
  opacity: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 .Overlay--before-close {
  opacity: 0;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c10 >:not(:last-child) {
  margin-right: 1em;
}

.c12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  color: var(--color-text);
  background-color: var(--color-bg);
  padding: 0.75em 2em;
  border-radius: 4px;
  font-size: 1em;
  font-weight: 600;
  cursor: pointer;
  line-height: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  position: relative;
  overflow: hidden;
  border: 1px solid var(--color-bg);
  --color-text: #3B7A9E;
  --color-bg: #FFFFFF;
  border: 1px solid var(--color-text);
}

.c12[disabled] {
  pointer-events: none;
  opacity: 0.6;
}

.c12:focus-within {
  border-color: transparent;
  outline: 3px solid #FDBD56;
  box-shadow: 0 0 0 3px #FDBD56;
}

.c12:hover {
  --color-text: #FFFFFF;
  --color-bg: #3B7A9E;
  border-color: var(--color-bg);
}

.c13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  color: var(--color-text);
  background-color: var(--color-bg);
  padding: 0.75em 2em;
  border-radius: 4px;
  font-size: 1em;
  font-weight: 600;
  cursor: pointer;
  line-height: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  position: relative;
  overflow: hidden;
  border: 1px solid var(--color-bg);
  --color-text: #FFFFFF;
  --color-bg: #3B7A9E;
  position: relative;
  border: none;
}

.c13[disabled] {
  pointer-events: none;
  opacity: 0.6;
}

.c13:focus-within {
  border-color: transparent;
  outline: 3px solid #FDBD56;
  box-shadow: 0 0 0 3px #FDBD56;
}

.c13:hover {
  --color-text: #FFFFFF;
  --color-bg: #2d5d79;
  border-color: var(--color-bg);
}

.c5 .Modal {
  max-width: 500px;
}

.c9 {
  color: #666666;
}

.c11 {
  padding-top: 1.5em;
}

<body
    aria-hidden="true"
    class="ReactModal__Body--open"
  >
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div
      id="toast"
    >
      <div
        class="c0"
      >
        <div
          class="c1"
        >
          <div
            class="c2"
          >
            <ol
              aria-live="assertive"
              aria-relevant="additions removals"
              class="c3"
              role="alert"
            />
          </div>
        </div>
      </div>
    </div>
    <div />
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div
      class="c4 c5"
    >
      <div
        class="Overlay Overlay--after-open"
      >
        <div
          class="Modal Modal--after-open"
          data-test="paste-modal"
          role="dialog"
          tabindex="-1"
        >
          <div
            class="c6 c7"
          >
            <button
              aria-label="Close"
              class="c8"
              role="button"
            >
              Ã—
            </button>
            <h2
              class="c9"
            >
              Confirm the removal of extra spaces from copied content
            </h2>
            <div
              class=""
            >
              <p>
                The copied content contains extra spaces at the start of lines of text, between words, or at the end of lines of text.
              </p>
              <p>
                Extra spaces need to be removed before this content can be pasted. Confirming will remove them automatically, while cancelling will prevent pasting.
              </p>
            </div>
            <div
              class="c10 c11"
              data-test="button-group"
            >
              <button
                class="c12"
                data-test="paste-modal-cancel"
                type="button"
              >
                Cancel
              </button>
              <button
                class="c13"
                data-test="paste-modal-confirm"
                type="button"
              >
                Confirm
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "history": Object {
    "action": "POP",
    "block": [Function],
    "canGo": [Function],
    "createHref": [Function],
    "entries": Array [
      Object {
        "hash": "",
        "key": "6i6lea",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    ],
    "go": [Function],
    "goBack": [Function],
    "goForward": [Function],
    "index": 0,
    "length": 1,
    "listen": [Function],
    "location": Object {
      "hash": "",
      "key": "6i6lea",
      "pathname": "/",
      "search": "",
      "state": undefined,
    },
    "push": [Function],
    "replace": [Function],
  },
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
